File: backend/internal/render/markdown.go
package render

import (
	"fmt"
	"strings"

	"github.com/ashley/drama-workbench/internal/model"
)

func Markdown(p *model.Plan) string {
	var b strings.Builder
	w := func(f string, a ...any) { fmt.Fprintf(&b, f, a...) }
	w("# %s\n\n", nz(p.Bible.Title, "Untitled Series"))
	w("## Concept\n- **Logline:** %s\n- **Theme:** %s\n- **Audience:** %s\n- **Tone:** %s\n- **Payoff engine:** %s\n- **Core conflict:** %s\n\n",
		p.Concept.Logline, p.Concept.Theme, p.Concept.Audience, p.Concept.Tone, p.Concept.PayoffEngine, p.Concept.CoreConflict)
	w("## Series Bible\n- Episodes: %d x %ds | Platform: %s\n- Integration: %s\n\n",
		p.Bible.Episodes, p.Bible.EpisodeSecs, p.Bible.Platform, p.Bible.IntegrationThesis)
	w("## Characters\n")
	for _, c := range p.Characters {
		w("- **%s** (%s): %s _Arc:_ %s\n", c.Name, c.Role, c.Bio, c.Arc)
	}
	w("\n## Episodes\n")
	for _, e := range p.Episodes {
		w("### Ep %d — %s\n%s\n- **Hook:** %s\n- **Cliffhanger:** %s\n- **Payoff:** %s\n\n",
			e.Number, e.Title, e.Synopsis, e.Hook, e.Cliffhanger, e.Payoff)
	}
	w("## Brand Integration\n")
	for _, pl := range p.Placements {
		w("- Ep %d — %s (%s): %s | CTA: %s\n", pl.Episode, pl.Category, pl.ProductSKU, pl.EmotionalBeat, pl.CTATiming)
	}
	w("\n## Hero Scenes\n")
	for _, h := range p.HeroScenes {
		w("### Ep %d — %s\n", h.Episode, h.Title)
		for _, s := range h.Shots {
			w("%d. [%s] %s — \"%s\"\n", s.Number, s.ShotType, s.Action, s.Dialogue)
		}
	}
	w("\n## Production\n- Format: %s | Budget: %s | Shots: %d | Cast: %d\n- Furniture: %s\n\n",
		p.Production.Format, p.Production.BudgetTier, p.Production.ShotCount, p.Production.CastSize, strings.Join(p.Production.FurnitureProps, ", "))
	w("## Distribution\n- CTA: %s\n- Hashtags: %s\n", p.Distribution.CTACopy, strings.Join(p.Distribution.Hashtags, " "))
	return b.String()
}

func nz(s, d string) string {
	if strings.TrimSpace(s) == "" {
		return d
	}
	return s
}

File: frontend/lib/markdown.ts
import { Plan } from "./types"

// planToMarkdown mirrors the Go backend's render.Markdown so the client can
// produce an identical Markdown export without a server round-trip.
const nz = (s: string, d: string) => (s && s.trim() !== "" ? s : d)

export function planToMarkdown(p: Plan): string {
  const out: string[] = []
  const w = (s: string) => out.push(s)

  w(`# ${nz(p.bible?.title ?? "", "Untitled Series")}\n`)

  const c = p.concept ?? ({} as Plan["concept"])
  w("## Concept")
  w(`- **Logline:** ${c.logline ?? ""}`)
  w(`- **Theme:** ${c.theme ?? ""}`)
  w(`- **Audience:** ${c.audience ?? ""}`)
  w(`- **Tone:** ${c.tone ?? ""}`)
  w(`- **Payoff engine:** ${c.payoffEngine ?? ""}`)
  w(`- **Core conflict:** ${c.coreConflict ?? ""}\n`)

  const b = p.bible ?? ({} as Plan["bible"])
  w("## Series Bible")
  w(`- Episodes: ${b.episodes ?? 0} x ${b.episodeSecs ?? 0}s | Platform: ${b.platform ?? ""}`)
  w(`- Integration: ${b.integrationThesis ?? ""}\n`)

  w("## Characters")
  for (const ch of p.characters ?? []) {
    w(`- **${ch.name}** (${ch.role}): ${ch.bio} _Arc:_ ${ch.arc}`)
  }
  w("")

  w("## Episodes")
  for (const e of p.episodes ?? []) {
    w(`### Ep ${e.number} — ${e.title}`)
    w(e.synopsis ?? "")
    w(`- **Hook:** ${e.hook ?? ""}`)
    w(`- **Cliffhanger:** ${e.cliffhanger ?? ""}`)
    w(`- **Payoff:** ${e.payoff ?? ""}\n`)
  }

  w("## Brand Integration")
  for (const pl of p.placements ?? []) {
    w(`- Ep ${pl.episode} — ${pl.category} (${pl.productSku}): ${pl.emotionalBeat} | CTA: ${pl.ctaTiming}`)
  }
  w("")

  w("## Hero Scenes")
  for (const h of p.heroScenes ?? []) {
    w(`### Ep ${h.episode} — ${h.title}`)
    for (const s of h.shots ?? []) {
      w(`${s.number}. [${s.shotType}] ${s.action} — "${s.dialogue}"`)
    }
  }
  w("")

  const pr = p.production ?? ({} as Plan["production"])
  w("## Production")
  w(`- Format: ${pr.format ?? ""} | Budget: ${pr.budgetTier ?? ""} | Shots: ${pr.shotCount ?? 0} | Cast: ${pr.castSize ?? 0}`)
  w(`- Furniture: ${(pr.furnitureProps ?? []).join(", ")}\n`)

  const d = p.distribution ?? ({} as Plan["distribution"])
  w("## Distribution")
  w(`- CTA: ${d.ctaCopy ?? ""}`)
  w(`- Hashtags: ${(d.hashtags ?? []).join(" ")}`)

  return out.join("\n")
}
